migrations/20210416143836-add-kind-to-transactions.js

```diff
@@ -0,0 +1,74 @@
+'use strict';
+
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    // 1. Add the 'kind' column on Transactions table
+    await queryInterface.addColumn('Transactions', 'kind', {
+      allowNull: true,
+      type: Sequelize.ENUM(
+        'ADDED_FUNDS',
+        'CONTRIBUTION',
+        'EXPENSE',
+        'HOST_FEE',
+        'HOST_FEE_SHARE',
+        'PAYMENT_PROCESSOR_FEE',
+        'PLATFORM_FEE',
+        'PLATFORM_TIP',
+        'PREPAID_PAYMENT_METHOD',
+      ),
+    });
+
+    // 2. Pre-fill the 'kind' columns everywhere we can
+    // ADDED_FUNDS
+    await queryInterface.sequelize.query(`
+      UPDATE "Transactions" t
+      SET "kind" = 'ADDED_FUNDS'
+      FROM "PaymentMethods" pm
+      WHERE t."PaymentMethodId" = pm.id
+      AND pm.service = 'opencollective'
+      AND pm."type" = 'host'
+    `);
+
+    // CONTRIBUTION
+    // This will set KIND=CONTRIBUTION on all Transactions with an order, except for the ones
+    // we already migrated above (added funds) and prepaid budgets (tier #15445).
+    await queryInterface.sequelize.query(`
+      UPDATE "Transactions" t
+      SET "kind" = 'CONTRIBUTION'
+      FROM "Orders" o
+      WHERE t."OrderId" IS NOT NULL
+      AND t."OrderId" = o.id
+      AND o."TierId" != 15445
+      AND "kind" IS NULL
+    `);
+
+    // PREPAID_PAYMENT_METHOD
+    // It's a bit ugly to use the `description` for that, but prod data shows that it only returns
+    // correct entries.
+    await queryInterface.sequelize.query(`
+      UPDATE "Transactions" 
+      SET "kind" = 'PREPAID_PAYMENT_METHOD'
+      WHERE description = 'Prepaid Budget'
+      OR description = 'Prepaid Payment Method for Gift Card Budget'
+    `);
+
+    // EXPENSE
+    await queryInterface.sequelize.query(`
+      UPDATE "Transactions" 
+      SET "kind" = 'EXPENSE'
+      WHERE "ExpenseId" IS NOT NULL
+    `);
+
+    // PLATFORM_TIP
+    await queryInterface.sequelize.query(`
+      UPDATE "Transactions"
+      SET "kind" = 'PLATFORM_TIP'
+      WHERE ("FromCollectiveId" = 8686 OR "CollectiveId" = 8686)
+      AND ("data" ->> 'isFeesOnTop')::boolean = TRUE
```

Comment:
I noticed that Transactions `388057` and `388056` are matching, maybe this is not accurate.

Can't we look at `PlatformTipForTransactionGroup` ?

Reply:
`388057` and `388056` seem like invalid transactions that we should fix. The order (#73185) is directed to Open Collective directly, so there's no reason to have `isFeesOnTop` in there (unless I'm missing something). I did not find similar occurrences in the pledged.

I'm afraid that we can miss some transactions if we look at `PlatformTipForTransactionGroup`, the following query returns a lot of results:
```sql
SELECT * 
FROM "Transactions" t 
WHERE ("FromCollectiveId" = 8686 OR "CollectiveId" = 8686)
AND ("data" ->> 'isFeesOnTop')::boolean = TRUE
AND "PlatformTipForTransactionGroup" IS NULL
```

Reply:
Seems `isFeesOnTop` means mainly that Platform Tips were attached to the original transaction. Should have been named `hasPlatformTips` and should not have been propagated to the Platform Tip transactions themselves.

Anyway, in this context I think this doing the job, except for the transaction that I highlighted.

Reply:
I've manually updated `388057` and `388056`, so the migration should be safe to run.

I'm still a bit concerned by the query above, there are basically two cases where the `PlatformTipForTransactionGroup` is `null` for valid platform tips:
- Refunds (not a big deal in this case)
- Transactions created before `2020-10-15 18:55:09` (see the query below). My guess is that https://github.com/opencollective/opencollective-api/blob/5654a97eb4417b80ebb90aca07931f0f8961678d/migrations/20201016085526-add-platform-contribution-transaction-group.js#L34 didn't catch everything. I suspect the join on the date to be the root cause for this. I initially proposed [another migration](https://github.com/opencollective/opencollective-api/pull/4716#pullrequestreview-512415868) joining on the payment providers payment IDs that seems to catch these cases. I'll try fixing that with https://github.com/opencollective/opencollective/issues/4205

```sql
SELECT *
FROM "Transactions" t 
INNER JOIN "Orders" o ON t."OrderId" = o.id
WHERE (t."FromCollectiveId" = 8686 OR t."CollectiveId" = 8686)
AND o."CollectiveId" != 8686
AND (t."data" ->> 'isFeesOnTop')::boolean = TRUE
AND t."PlatformTipForTransactionGroup" IS NULL
AND t."isRefund" IS NOT TRUE
ORDER BY t."createdAt" DESC
```